### PR TITLE
metadata: default implementation of MetaRegistry

### PR DESCRIPTION
**Pull request by xingjigongsi**
model

### CODE REVIEW

File: model_test.go [hunk 1]
```diff
@@ -36,8 +37,24 @@ func TestTagMetaRegistry(t *testing.T) {
 	assert.Equal(t, "id", idMeta.columnName)
 	assert.Equal(t, "Id", idMeta.fieldName)
 	assert.Equal(t, reflect.TypeOf(int64(0)), idMeta.typ)
-	assert.True(t, idMeta.isAutoIncrement)
-	assert.True(t, idMeta.isPrimaryKey)
+	assert.False(t, idMeta.isAutoIncrement)
+	assert.False(t, idMeta.isPrimaryKey)
 
 	// 把剩下的三个字段都断言一遍，注意它们的类型，而且它们也不是主键，也不是自增
```
Comment (flycash) [inline model_test.go:43]: remove this comment.

File: model_test.go [hunk 2]
```diff
@@ -36,8 +37,24 @@ func TestTagMetaRegistry(t *testing.T) {
...
+
+	idMetaFistName := meta.fieldMap["FistName"]
```
Comment (flycash) [inline model_test.go:45]: firstNameMeta

File: model_test.go [hunk 3]
```diff
@@ -36,8 +37,24 @@ func TestTagMetaRegistry(t *testing.T) {
...
+	assert.Equal(t, "fist_name", idMetaFistName.columnName)
+	assert.Equal(t, "FistName", idMetaFistName.fieldName)
+	assert.Equal(t, reflect.TypeOf(string("")), idMetaFistName.typ)
+
+	idMetaLastName := meta.fieldMap["LastName"]
```
Comment (flycash) [inline model_test.go:50]: lastNameMeta

File: model_test.go [hunk 4]
```diff
@@ -36,8 +37,24 @@ func TestTagMetaRegistry(t *testing.T) {
...
+	assert.Equal(t, "last_name", idMetaLastName.columnName)
+	assert.Equal(t, "LastName", idMetaLastName.fieldName)
+	assert.Equal(t, reflect.TypeOf(string("")), idMetaLastName.typ)
+
+	idMetaLastAge := meta.fieldMap["Age"]
```
Comment (flycash) [inline model_test.go:55]: ageMeta

File: model.go [hunk 5]
```diff
@@ -11,59 +11,85 @@
...
 // limitations under the License.
-
 package eql
 
 import (
 	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/gotomicro/eql/internal"
```
Comment (flycash) [inline model.go:21]: you need to commit this file

File: model.go [hunk 6]
```diff
@@ -11,59 +11,85 @@
...
+	columnName      string
+	fieldName       string
+	typ             reflect.Type
+	isPrimaryKey    bool
 	isAutoIncrement bool
 }
 
 type tableMetaOption func(meta *TableMeta)
 
 type MetaRegistry interface {
```
Comment (flycash) [inline model.go:43]: Write comments for it

File: model.go [hunk 7]
```diff
@@ -11,59 +11,85 @@
...
-	// 这里传过来的 table 应该是结构体指针，例如 *User
 	Get(table interface{}) (*TableMeta, error)
```
Comment (flycash) [inline model.go:44]: write comments for these two functions

File: model.go [hunk 8]
```diff
@@ -11,59 +11,85 @@
...
-	// 遍历 table 的所有字段
-	// 对每一个字段检查 tag eql, 如果 eql 里面有 auto_increment, primary_key，就相应设置 ColumnMeta的值
-	// 最后处理所有的 opts
-	// 塞到 map 里面
-	panic("implement me")
-}
+	rtype := reflect.TypeOf(table)
+	v := rtype.Elem()
+	columnMetas := []*ColumnMeta{}
+	lens := v.NumField()
```
Comment (flycash) [inline model.go:66]: renamed to fieldNum or numField

File: model_test.go (hunk 1 above)
Comment (xingjigongsi) [reply model_test.go:43]: ok

File: model_test.go (hunk 2 above)
Comment (xingjigongsi) [reply model_test.go:45]: ok

File: model_test.go (hunk 3 above)
Comment (xingjigongsi) [reply model_test.go:50]: ok

File: model_test.go (hunk 4 above)
Comment (xingjigongsi) [reply model_test.go:55]: ok

File: model.go (hunk 5 above)
Comment (xingjigongsi) [reply model.go:21]: ok

File: model.go (hunk 7 above)
Comment (xingjigongsi) [reply model.go:44]: ok

File: model.go (hunk 8 above)
Comment (xingjigongsi) [reply model.go:66]: ok

File: model.go [hunk 9]
```diff
@@ -11,59 +11,92 @@
...
+			columnName:      internal.UnderscoreName(structField.Name),
+			fieldName:       structField.Name,
+			typ:             structField.Type,
+			isAutoIncrement: isAuto,
+			isPrimaryKey:    isKey,
+		}
+		columnMetas = append(columnMetas, columnMeta)
+		fieldMap[columnMeta.fieldName] = columnMeta
+	}
+	TableMeta := &TableMeta{
```
Comment (flycash) [inline model.go:90]: local variable should start with lower case

File: model_test.go [hunk 10]
```diff
@@ -26,7 +27,7 @@ func TestTagMetaRegistry(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	assert.NotNil(t, err)
+	assert.Nil(t, err)
```
Comment (flycash) [inline model_test.go:30]: we already check the error, so we don't need to check it again

File: model_test.go (hunk 1 above)
Comment (flycash) [reply model_test.go:41]: remove the comment

File: model.go [hunk 11]
```diff
@@ -11,59 +11,92 @@
...
+
+	if v, ok := t.metas.Load(reflect.TypeOf(table)); ok {
+		return v.(*TableMeta), nil
+	}
+	return t.Register(table)
 }
 
+//Generate a metadata for each column and place it in a thread-safe map for easy access to the metadata directly.
+//And the metadata can be modified by user-defined methods
 func (t *tagMetaRegistry) Register(table interface{}, opts ...tableMetaOption) (*TableMeta, error) {
```
[the remark this answers is not in the document]
Comment (xingjigongsi) [reply model.go:69]: 嗯嗯

File: model.go (hunk 9 above)
Comment (xingjigongsi) [reply model.go:90]: ok

File: model.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (xingjigongsi) [reply model.go:69]: ok

**Review by flycash — APPROVED** [review]

**Pull request merged**